File: convex/functions/relations.ts
import { query } from "../_generated/server";
import { v } from "convex/values";
import { validateTableAccess, validateWhereClause, validateQueryOptions } from "../lib/validators";

/**
 * Relationship Helpers for CFX-DB
 *
 * Provides JOIN-like functionality to query related data across tables.
 * Fetches primary data first, then fetches related data using foreign keys.
 *
 * OPTIMIZED: Now uses limits and fetches related records more efficiently.
 */

// Maximum relation depth to prevent DoS
const MAX_RELATION_DEPTH = 3;
// Maximum records to fetch per relation to prevent memory issues
const MAX_RELATED_RECORDS = 1000;
// Maximum primary records
const MAX_PRIMARY_RECORDS = 500;

/**
 * Query with relations (JOIN-like)
 *
 * Fetches primary records and then fetches related data from other tables.
 *
 * PERFORMANCE NOTE: For best performance, ensure related tables have indexes
 * on their foreign key fields (e.g., vehicles should have "by_player_id" index).
 */
export const queryWithRelations = query({
    args: {
        table: v.string(),
        where: v.optional(v.any()),
        options: v.optional(
            v.object({
                orderBy: v.optional(v.any()),
                limit: v.optional(v.number()),
                offset: v.optional(v.number()),
                select: v.optional(v.array(v.string())),
            })
        ),
        relations: v.array(
            v.object({
                name: v.string(), // Alias for the relation
                table: v.string(), // Related table
                foreignKey: v.string(), // Field in related table
                localKey: v.optional(v.string()), // Field in this table (defaults to _id)
                type: v.union(v.literal("one"), v.literal("many")), // 1:1 or 1:N
                select: v.optional(v.array(v.string())), // Fields to include from related table
                indexName: v.optional(v.string()), // Optional: index name for faster lookups
                limit: v.optional(v.number()), // Max related records per parent (for "many" type)
            })
        ),
    },
    handler: async (ctx, args) => {
        // Validate table access
        validateTableAccess(args.table);

        // Validate where clause
        if (args.where) {
            validateWhereClause(args.where);
        }

        // Validate query options
        if (args.options) {
            validateQueryOptions(args.options);
        }

        // Validate relations
        if (args.relations.length > MAX_RELATION_DEPTH) {
            throw new Error(
                `Too many relations (max ${MAX_RELATION_DEPTH})`
            );
        }

        for (const relation of args.relations) {
            validateTableAccess(relation.table);
        }

        // Calculate effective limit for primary records
        const requestedLimit = args.options?.limit || MAX_PRIMARY_RECORDS;
        const effectiveLimit = Math.min(requestedLimit, MAX_PRIMARY_RECORDS);

        // Fetch primary records with limit
        let dbQuery = ctx.db.query(args.table as any);

        // Apply where clause
        if (args.where) {
            dbQuery = dbQuery.filter((q: any) => {
                let result: any = null;
                for (const [key, value] of Object.entries(args.where)) {
                    const condition = q.eq(q.field(key), value);
                    result = result === null ? condition : q.and(result, condition);
                }
                return result;
            });
        }

        // Fetch with limit + offset buffer
        const fetchLimit = effectiveLimit + (args.options?.offset || 0);
        let records = await dbQuery.take(Math.min(fetchLimit, MAX_PRIMARY_RECORDS));

        // Apply offset
        if (args.options?.offset) {
            records = records.slice(args.options.offset);
        }

        // Apply limit
        records = records.slice(0, effectiveLimit);

        // Apply field selection to primary records (but keep localKey for relations)
        const localKeysNeeded = new Set(args.relations.map(r => r.localKey || "_id"));
        if (args.options?.select) {
            records = records.map((r: any) => {
                const selected: any = { _id: r._id };
                for (const field of args.options!.select!) {
                    if (field in r) {
                        selected[field] = r[field];
                    }
                }
                // Keep localKeys even if not in select
                for (const key of localKeysNeeded) {
                    if (key in r && !(key in selected)) {
                        selected[key] = r[key];
                    }
                }
                return selected;
            });
        }

        // Fetch related data for each relation
        for (const relation of args.relations) {
            const localKey = relation.localKey || "_id";
            const localKeys = records.map((r: any) => r[localKey]);
            const uniqueLocalKeys = [...new Set(localKeys)];

            // Skip if no records to relate
            if (uniqueLocalKeys.length === 0) {
                for (const record of records) {
                    (record as any)[relation.name] = relation.type === "one" ? null : [];
                }
                continue;
            }

            // Fetch related records - use index if specified, otherwise limited fetch
            let relatedRecords: any[];

            if (relation.indexName) {
                // Use specified index for each unique key (more efficient for small key sets)
                const relatedPromises = uniqueLocalKeys.map(async (keyValue) => {
                    return ctx.db
                        .query(relation.table as any)
                        .withIndex(relation.indexName as any, (q: any) =>
                            q.eq(relation.foreignKey, keyValue)
                        )
                        .take(relation.limit || (relation.type === "one" ? 1 : 100));
                });
                const relatedArrays = await Promise.all(relatedPromises);
                relatedRecords = relatedArrays.flat();
            } else {
                // Fallback: fetch limited records and filter (less efficient)
                const allRelated = await ctx.db
                    .query(relation.table as any)
                    .take(MAX_RELATED_RECORDS);

                relatedRecords = allRelated.filter((r: any) =>
                    uniqueLocalKeys.includes(r[relation.foreignKey])
                );
            }

            // Apply field selection to related records
            const processedRelated = relation.select
                ? relatedRecords.map((r: any) => {
                      const selected: any = { _id: r._id };
                      // Always include foreign key for matching
                      selected[relation.foreignKey] = r[relation.foreignKey];
                      for (const field of relation.select!) {
                          if (field in r) {
                              selected[field] = r[field];
                          }
                      }
                      return selected;
                  })
                : relatedRecords;

            // Attach related data to primary records
            for (const record of records) {
                const localValue = (record as any)[localKey];

                if (relation.type === "one") {
                    (record as any)[relation.name] =
                        processedRelated.find(
                            (r: any) => r[relation.foreignKey] === localValue
                        ) || null;
                } else {
                    const related = processedRelated.filter(
                        (r: any) => r[relation.foreignKey] === localValue
                    );
                    // Apply per-parent limit for "many" relations
                    (record as any)[relation.name] = relation.limit
                        ? related.slice(0, relation.limit)
                        : related;
                }
            }
        }

        return records;
    },
});

/**
 * Nested query with relations
 *
 * Allows fetching relations that themselves have relations (depth: 2+)
 * Use with caution - can be slow with deep nesting.
 */
export const queryWithNestedRelations = query({
    args: {
        table: v.string(),
        where: v.optional(v.any()),
        relations: v.array(
            v.object({
                name: v.string(),
                table: v.string(),
                foreignKey: v.string(),
                localKey: v.optional(v.string()),
                type: v.union(v.literal("one"), v.literal("many")),
                select: v.optional(v.array(v.string())),
                // Nested relations
                relations: v.optional(v.array(v.any())),
            })
        ),
    },
    handler: async (ctx, args) => {
        // TODO: Implement nested relations
        // For now, just call queryWithRelations without nesting
        throw new Error("Nested relations not yet implemented - use queryWithRelations for now");
    },
});

/**
 * Count with relations
 *
 * Useful for getting counts of related data without fetching all records.
 *
 * OPTIMIZED: Uses limits and index-based lookups when available.
 * For high-performance counting at scale, consider using Convex Aggregate component.
 */
export const countWithRelations = query({
    args: {
        table: v.string(),
        where: v.optional(v.any()),
        relation: v.object({
            table: v.string(),
            foreignKey: v.string(),
            localKey: v.optional(v.string()),
            indexName: v.optional(v.string()), // Optional: index for faster counting
        }),
        limit: v.optional(v.number()), // Max primary records to count for
    },
    handler: async (ctx, args) => {
        validateTableAccess(args.table);
        validateTableAccess(args.relation.table);

        const maxRecords = Math.min(args.limit || 100, MAX_PRIMARY_RECORDS);

        // Fetch primary records with limit
        let dbQuery = ctx.db.query(args.table as any);

        if (args.where) {
            dbQuery = dbQuery.filter((q: any) => {
                let result: any = null;
                for (const [key, value] of Object.entries(args.where)) {
                    const condition = q.eq(q.field(key), value);
                    result = result === null ? condition : q.and(result, condition);
                }
                return result;
            });
        }

        const records = await dbQuery.take(maxRecords);

        // Count related records for each primary record
        const localKey = args.relation.localKey || "_id";
        const counts: { [key: string]: number } = {};

        if (args.relation.indexName) {
            // Use index for efficient counting
            const countPromises = records.map(async (record: any) => {
                const localValue = record[localKey];
                const related = await ctx.db
                    .query(args.relation.table as any)
                    .withIndex(args.relation.indexName as any, (q: any) =>
                        q.eq(args.relation.foreignKey, localValue)
                    )
                    .take(10000); // Cap at reasonable limit
                return { key: localValue, count: related.length };
            });
            const results = await Promise.all(countPromises);
            for (const { key, count } of results) {
                counts[key] = count;
            }
        } else {
            // Fallback: fetch limited related records
            const allRelated = await ctx.db
                .query(args.relation.table as any)
                .take(MAX_RELATED_RECORDS);

            for (const record of records) {
                const localValue = (record as any)[localKey];
                const count = allRelated.filter(
                    (r: any) => r[args.relation.foreignKey] === localValue
                ).length;
                counts[localValue] = count;
            }
        }

        return counts;
    },
});

File: lua/relations.lua
--[[
    Relationship Query API for CFX-DB

    Provides JOIN-like functionality to query related data across tables.

    Example usage:
        DB.relations('players', {
            where = {identifier = license},
            relations = {
                {
                    name = 'vehicles',
                    table = 'vehicles',
                    foreignKey = 'owner',
                    localKey = 'identifier',
                    type = 'many'
                },
                {
                    name = 'inventory',
                    table = 'inventory',
                    foreignKey = 'owner',
                    localKey = 'identifier',
                    type = 'many',
                    select = {'item', 'count'}
                }
            }
        }, function(success, players)
            if success then
                local player = players[1]
                print("Player has", #player.vehicles, "vehicles")
                print("Player has", #player.inventory, "items")
            end
        end)
]]

local Relations = {}

-- Detect which cfx-db resource is running (dev: cfx-db-dev, prod: cfx-db)
local cfxDbResource = GetResourceState('cfx-db-dev') == 'started' and 'cfx-db-dev' or 'cfx-db'

---Query with relations (JOIN-like)
---@param table string Primary table name
---@param params table Query parameters {where, options, relations}
---@param callback function Callback function(success, records)
function Relations.query(table, params, callback)
    if type(table) ~= "string" then
        error("Relations.query: table must be a string")
    end

    if type(params) ~= "table" then
        error("Relations.query: params must be a table")
    end

    if type(callback) ~= "function" then
        error("Relations.query: callback must be a function")
    end

    if not params.relations or type(params.relations) ~= "table" then
        error("Relations.query: params.relations must be a table")
    end

    -- Validate relations
    for i, relation in ipairs(params.relations) do
        if not relation.name then
            error(string.format("Relations.query: relation %d missing 'name' field", i))
        end

        if not relation.table then
            error(string.format("Relations.query: relation %d missing 'table' field", i))
        end

        if not relation.foreignKey then
            error(string.format("Relations.query: relation %d missing 'foreignKey' field", i))
        end

        if not relation.type or (relation.type ~= "one" and relation.type ~= "many") then
            error(string.format("Relations.query: relation %d 'type' must be 'one' or 'many'", i))
        end
    end

    local args = {
        table = table,
        where = params.where,
        options = params.options,
        relations = params.relations
    }

    exports[cfxDbResource]:query('relations.queryWithRelations', args, callback)
end

---Query with relations (async/await style)
---@param table string Primary table name
---@param params table Query parameters {where, options, relations}
---@return boolean success
---@return any records
function Relations.queryAsync(table, params)
    if type(table) ~= "string" then
        error("Relations.queryAsync: table must be a string")
    end

    if type(params) ~= "table" then
        error("Relations.queryAsync: params must be a table")
    end

    if not params.relations or type(params.relations) ~= "table" then
        error("Relations.queryAsync: params.relations must be a table")
    end

    local args = {
        table = table,
        where = params.where,
        options = params.options,
        relations = params.relations
    }

    local success, result = exports[cfxDbResource]:queryAsync('relations.queryWithRelations', args)
    return success, result
end

---Count related records
---@param table string Primary table name
---@param params table Query parameters {where, relation}
---@param callback function Callback function(success, counts)
function Relations.count(table, params, callback)
    if type(table) ~= "string" then
        error("Relations.count: table must be a string")
    end

    if type(params) ~= "table" then
        error("Relations.count: params must be a table")
    end

    if type(callback) ~= "function" then
        error("Relations.count: callback must be a function")
    end

    if not params.relation or type(params.relation) ~= "table" then
        error("Relations.count: params.relation must be a table")
    end

    local args = {
        table = table,
        where = params.where,
        relation = params.relation
    }

    exports[cfxDbResource]:query('relations.countWithRelations', args, callback)
end

-- Helper: Simple one-to-many query
---Fetch primary record with related records
---@param primaryTable string Primary table name
---@param primaryWhere table Where clause for primary table
---@param relationName string Name for the relation in results
---@param relationTable string Related table name
---@param foreignKey string Foreign key in related table
---@param localKey string|nil Local key in primary table (default: identifier or _id)
---@param callback function Callback function(success, record)
function Relations.oneToMany(primaryTable, primaryWhere, relationName, relationTable, foreignKey, localKey, callback)
    -- Handle optional localKey parameter
    if type(localKey) == "function" then
        callback = localKey
        localKey = nil
    end

    Relations.query(primaryTable, {
        where = primaryWhere,
        relations = {
            {
                name = relationName,
                table = relationTable,
                foreignKey = foreignKey,
                localKey = localKey,
                type = "many"
            }
        }
    }, function(success, records)
        if success and records and #records > 0 then
            callback(true, records[1])
        else
            callback(success, records)
        end
    end)
end

-- Helper: Simple one-to-one query
---Fetch primary record with a single related record
---@param primaryTable string Primary table name
---@param primaryWhere table Where clause for primary table
---@param relationName string Name for the relation in results
---@param relationTable string Related table name
---@param foreignKey string Foreign key in related table
---@param localKey string|nil Local key in primary table
---@param callback function Callback function(success, record)
function Relations.oneToOne(primaryTable, primaryWhere, relationName, relationTable, foreignKey, localKey, callback)
    -- Handle optional localKey parameter
    if type(localKey) == "function" then
        callback = localKey
        localKey = nil
    end

    Relations.query(primaryTable, {
        where = primaryWhere,
        relations = {
            {
                name = relationName,
                table = relationTable,
                foreignKey = foreignKey,
                localKey = localKey,
                type = "one"
            }
        }
    }, function(success, records)
        if success and records and #records > 0 then
            callback(true, records[1])
        else
            callback(success, records)
        end
    end)
end

-- Expose relations API
_G.DBRelations = Relations

-- Add to main DB namespace
if DB then
    DB.relations = Relations.query
    DB.relationsAsync = Relations.queryAsync
    DB.relationsCount = Relations.count
    DB.oneToMany = Relations.oneToMany
    DB.oneToOne = Relations.oneToOne
end

return Relations
